오픈마켓 [Step3] 산체스

sanches37 opened the pull request:
@GREENOVER 

안녕하세요 그린! 마지막날에 되서야 step3 PR을 보냅니다.
고민을 많이 했지만, 진전이 많이없었어요. 구현한 부분만이라도 리뷰를 받아보고 싶어서 PR을 보냅니다.

##고민한점
1. 포토앨범을 멀티픽으로 불러오는 기능을 IOS14 이상이나 오픈라이브러리를 사용해지 않고 구현하기 위해 커스텀 포토앨범을 만들었습니다.
많이 찾아보고 구현했지만 이러한 방식으로 구현하면 되는지 궁금합니다.

2 .등록/수정 화면을 하나의 collectionView 로 구현했습니다. 이유는 후에 수정이 이루어진다면 대응하기에 더 좋다고 판단해서입니다.
하지만 구현해본 결과 복잡한 작업들을 많이해야되서 step3을 완성하지 못한 원인이 된 것 같아요 ㅎㅎ;;
특히 textField 를 collectionView 로 구현했더니 값을 찾아서 작업하는데 너무 어려웠고 제대로 구현하지 못했습니다.

3. 등록/수정 화면의 처리를 protocol 을 최대한 사용하여 코드내 분기처리를 줄이고 싶었습니다.  이 부분에서도 고민을 많이했지만, 구현을 제대로 하진 못한 것 같아요. 이 부분과 2번 고민한 점은 PR을 보내고도 더 좋은 방법을 계속 고민할 것 같습니다.

4. step2 와 마찬가지로 step3 에서도 ViewController 와 CollectionView 를 분리해서 구현을 했습니다. 
ViewController 하나에서 처리하는게 얼마나 편하고 감사한 작업인지 알게 되었습니다. 
화면 전환이 빈번하게 일어나게 되서 Delegate 나 Closure 를 통한 데이터 전달 코드를 사용해 볼 수 있는 경험이 되었습니다. 
분리로 인해서 ViewController 가 아닌 타입에서 화면 전환을 해야되는 상황도 생겼는데,
 여러번 코드를 수정해서 다시 ViewController 에서 화면 전환 작업을 할 수 있도록 변경했습니다. 
 그 과정에서 뷰컨트롤러가 아닌 타입에서 굳이 화면 전환을 하려면 TopViewController 를 불러와서 처리할 수 있다는 걸 알게 되었습니다.
코드를 어느 타입에 배치할까도 고민할 수 있어서 ViewController 와 CollectionView 분리 적용은 좋은 선택이었던 것 같습니다.

step3을 진행하면서, 정해진 틀은 있지만 세부적으로 내가 생각한대로 구현해볼 수 있어서 직접 앱을 만들어 보는 것 같이 작업을 재밌게 했습니다.
첫 단추에 욕심을 부려서 어려운 방향으로 끌고가서 마무리를 제대로 짓진 못했지만, 그 과정에서 배운게 많다고 생각합니다.
첫 개인프로젝트였는데, 내가 해결하지 못하면 작업이 진행이 되지 않으니 한주동안 이것만 생각하면서 지냈던 것 같습니다.
 그동안 팀 프로젝트를 하면서 팀원들에게 너무 의존적이지 않았나 생각도 듭니다. 
3주동안 좋은 말씀많이 해주시고, 좋은 리뷰해주셔서 감사합니다. 그린에게 리뷰를 받을 수 있어서 영광이었습니다.
PR이 늦어서 리뷰는 천천히 해주셔도 될 것 같습니다. 추석 잘 보내세요~

sanches37 committed:
feat: 버튼 클릭시 등록 및 수정 얼럿 구현
OpenMarketViewController 내 enrollModifyButton 구현
버튼 클릭시 얼럿으로 등록, 수정 정할 수 있도록 구현

sanches37 committed:
feat: 수정, 등록에 따른 화면 전환 다르게 지정
storyboard segue 로 EnrollModifyViewController 연결
alert 으로 화면 전환시 등록, 수정에 맞게 이동하도록 구

sanches37 committed:
refactor: alert title 매직넘버 쓰지 않도록 수정

sanches37 committed:
feat: 등록/수정 화면 collecionView photoCell 구현
EnrollModifyCollectionViewDataSource, EnrollModifyCollectionViewDelegate 구현
EnrollModifyPhotoCell 타입 구현
EnrollModifyViewController 와 EnrollModifyCollectionViewDataSource 연결
EnrollModifyPhotoCell 와 OpenMarketCollectionViewDataSource identifier 로 연결

sanches37 committed:
feat: photocell Layout 지정

sanches37 committed:
test: textFied 배치를 위한 collectionView 시험 작업
EnrollModifyListCell 타입 구현
textField 시험용으로 하나의 collectionView 에 두가지 셀 작동 시험

sanches37 committed:
refactor: 레이아웃 중복 코드 줄이려 구조 수정
리스트 레이아웃 코드를 두번 쓰고 있어서 decidedListLayout 의 호출 위치 이동

sanches37 committed:
feat: 등록/수정 화면 textField 구현
EnrollModifyListCell textField 가 적용되도록 코드 구현
EnrollModifyCollectionViewDataSource 내 PlaceholderList 생성
PlaceholderList 로 collectionView 동작하도록 구현

sanches37 committed:
refactor: photo 선택버튼 collectionView에 추가
EnrollModifyPhotoSelectCell 타입 구현
EnrollModifyPhotoSelectCell 이 collectionView 에 적용되도록 코드 수정

sanches37 committed:
feat: 사진첩 이미지를 가져오기 위한 AppDelegate 코드작성
info.plist photo privacy 추가

sanches37 committed:
feat: photoAlbumViewController 타입 구현
cell Button 클릭시 PhotoAlbumViewController 로 이동하도록 코드 구현

sanches37 committed:
refactor: 포토앨범 접근 코드 이동
포토앨범 접근 코드 PhotoAlbumCollectionViewDataSource 로 이동

sanches37 committed:
feat: PhotoAlbum DataSource 코드 구현

sanches37 committed:
feat: photoAlBumViewController 레이아웃 적용
requestPhotoAlbum 기능 없애고 다시 AppDelegate 에서 동작하게 코드 수정

sanches37 committed:
feat: 포토앨범 선택/해제 기능 구현
PhotoAlbumCollectionViewDelegate 에서 사진 선택 해제 코드 추가
photoAlbum 으로 화면 전환 방식 Segue 로 변경

sanches37 committed:
refactor: 사진첩 Asset 불러오는 방식 변경
PhotoAlbumManager 로 사진첩의 이미지를 불러올 수 있게 구조 변경

sanches37 committed:
feat: 선택한 이미지를 전달할 selectPhotoAlbumImage 생성

sanches37 committed:
refactor: 포토앨범으로 화면 전환 코드 변경
UICollectionViewDelegate 를 이용해서 화면 전환 하도록 변경

sanches37 committed:
feat: 사진첩에서 사진선택시 선택이미지 표시되도록 구현
클로저를 이용하여 선택화면 이미지 EnrollModifyViewController 로 전달
선택화면의 delete 이미지 추가
EnrollModifyCell, PhotoAlbumCell 코드 정리

sanches37 committed:
feat: 사진첩에서 가져온 사진 삭제 코드 추가

sanches37 committed:
refactor: collectionViewDelegate 분리를 위해 코드 수정
photoSelectButton 을 EnrollModifyViewController 로 이동
EnrollModifyCollectionViewDelegate 타입 생성

sanches37 committed:
feat: 텍스트필드 입력 내용 네트워크 통신 구현
parameter 타입 내 입력 판단하는 judgeNil 메서드 구현
등록/수정을 구분하여 코드를 수행할 수 있게 프로토콜 구현
프로토콜에 필요한 enum 타입구현
textField 입력과 확인에 필요한 코드구현

sanches37 committed:
refactor: 같은 종류 코드 함수로 묶음처리 및 코드정리

sanches37 committed:
refactor: Parameter 타입에 id 프로퍼티 추가
id 프로퍼티 추가하여 enrollModifyViewController 에서 사용하도록 코드 구현

sanches37 committed:
chore: print 삭제

sanches37 committed:
refactor: id Nil 처리하여 통신에 사용하도록 코드 수정

sanches37 committed:
refactor: patch 주소 오기 수정

GREENOVER commented on OpenMarket/OpenMarket/AppDelegate.swift:
```diff
@@ -19,21 +18,28 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         return true
     }
-
+    
```
질문 있습니다🙋🏻‍♂️
요런 들여쓰기가 들어가는 부분들은 SwiftLint가 적용되어 컨벤션이 자동으로 맞춰준건가요👀

GREENOVER commented on OpenMarket/OpenMarket/AppDelegate.swift:
```diff
@@ -19,21 +18,28 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        
+        if !changeOrientation {
+            return [.portrait]
+        } else {
```
요부분은 else 조건문을 태우지 않고 바로 return문을 넣어주어도 earlyExit를 할 수 있습니다!
지금의 경우는 성능이나 그런 부분에서 차이가 별로 없지만?! 불필요한 코드도 줄고 가져가는 이점이 있을꺼에요☺️

GREENOVER commented on OpenMarket/OpenMarket/AppDelegate.swift:
```diff
@@ -8,9 +8,8 @@ import UIKit
-
-
-
+    var changeOrientation: Bool = true
```
AppDelegate에서 화면전환 처리를 해주는점 좋아요👍🏻

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,130 @@
+        2
+    }
+    
+    func collectionView(_ collectionView: UICollectionView,
+                        numberOfItemsInSection section: Int) -> Int {
```
처음보다 훨씬 파라미터 시그니쳐 가독성이 높아졌습니다! 멋져요💯🎉

GREENOVER replied:
하나 더 욕심 내보자면 
```swift
numberOfItemsInSection section: Int
```
이 부분의 시그니쳐가 길어진다면 -> Int로 시작하는 부분도 가독성이 좀 떨어질 수도 있을것 같아요!
저는 그래서 보통
```swift
func collectionView(
    _ collectionView: UICollectionView,
    numberOfItemsInSection section: Int
    ) -> Int {
    ... 
}
```
이런식으로 파라미터의 끝을 따라가지 않아도 리턴값을 좀 더 쉽게 파악할 수 있도록 컨벤션을 맞추고 있습니다!
이건 코드 스타일이라 산체스가 아.. 이런것도 있구나!라는 생각 정도 가지고 본인의 스타일에 맞는걸 택하면 좋을것 같아요😀

YongHoonJJo replied:
https://google.github.io/swift/#function-declarations

@sanches37 요기 참고해보면 좋을 것 같아요!! :)

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,130 @@
+        if section == .zero {
+            return photoAlbumImages.count + photoSelectButton.count
+        }
+        return placeholderList.count
```
이 코드를 보기전에 위에서 언급했었는데..! earlyExit가 잘 사용됐네요..! 크👏

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,130 @@
+    func collectionView(_ collectionView: UICollectionView,
+                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        if indexPath.item == .zero && indexPath.section == .zero {
+            guard let photoSelectCell =
+                    collectionView.dequeueReusableCell(withReuseIdentifier: EnrollModifyPhotoSeclectCell.identifier, for: indexPath) as? EnrollModifyPhotoSeclectCell else {
+                return UICollectionViewCell()
```
요부분이 대체적으로 호출메서드나 파라미터 인자들도 그렇고 정의하는것이 필수적으로 길어서 그럴 수 있지만..!? 조금 가독성이 떨어진다고 생각해요!
가독성을 좀 더 높여볼까요??📈

sanches37 committed:
refactor: 입력사항 확인 오류로 코드 수정
"" 로 받아진 String 을 nil 로 처리하지 않아 분기 구문 추가

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,130 @@
+    
+    func collectionView(_ collectionView: UICollectionView,
+                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        if indexPath.item == .zero && indexPath.section == .zero {
+            guard let photoSelectCell =
+                    collectionView.dequeueReusableCell(withReuseIdentifier: EnrollModifyPhotoSeclectCell.identifier, for: indexPath) as? EnrollModifyPhotoSeclectCell else {
+                return UICollectionViewCell()
+            }
+            let btn = photoSelectButton[indexPath.item]
+            photoSelectCell.configure(photoSelectButton: btn)
+            
+            return photoSelectCell
+        }
+        if indexPath.item != .zero && indexPath.section == .zero {
+            guard let photoCell =
+                    collectionView.dequeueReusableCell(withReuseIdentifier: EnrollModifyPhotoCell.identifier, for: indexPath) as? EnrollModifyPhotoCell else {
+                return UICollectionViewCell()
+            }
+            let photoAlbumImageForItem = photoAlbumImages[indexPath.item - 1]
+            photoCell.configure(image: photoAlbumImageForItem)
+            
+            return photoCell
+        } else {
+            guard let listCell =
+                    collectionView.dequeueReusableCell(withReuseIdentifier: EnrollModifyListCell.identifier, for: indexPath) as? EnrollModifyListCell else {
+                return UICollectionViewCell()
+            }
+            let placeholderForItem = placeholderList[indexPath.item]
+            listCell.configure(
+                placeholderList: placeholderForItem.description, dataSource: self)
+            
+            return listCell
+        }
```
지금같은 조건문을 통한 비교도 너무 좋습니다!
그치만 만약 이런 조건들이 무궁무진해진다면 코드 유지보수가 어려워질것 같아요😱
혹시 이런 방법은 어떨까요?
조건을 비교하되 cell 상수에 대한 선언부와 configure해주는 부분등 현재 공통되는 코드는 묶어주고
지금 변하는것들은 어떤셀을 재사용할지인데 이 부분은 Enum으로 만들어서 관리해도 될것같다는 생각이 듭니다!

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,130 @@
+    
+    func createCompositionalLayout() -> UICollectionViewCompositionalLayout {
+        return UICollectionViewCompositionalLayout { (sectionNumber, _) -> NSCollectionLayoutSection? in
+            switch sectionNumber {
+            case 0:
+                let phothCellMargin = self.compositionalLayout.margin(
+                    top: 5, leading: 0, bottom: 5, trailing: 5)
+                let photoViewMargin = self.compositionalLayout.margin(
+                    top: 0, leading: 5, bottom: 0, trailing: 0)
+                return self.compositionalLayout.enrollLayout(
+                    portraitHorizontalNumber: 4,
+                    landscapeHorizontalNumber: 4,
+                    cellVerticalSize: .fractionalHeight(1/6),
+                    scrollDirection: .horizontal,
+                    cellMargin: phothCellMargin,
+                    viewMargin: photoViewMargin)
+            default:
+                let listCellMargin = self.compositionalLayout.margin(
+                    top: 0, leading: 0, bottom: 0, trailing: 5)
+                let listViewMargin = self.compositionalLayout.margin(
+                    top: 0, leading: 5, bottom: 0, trailing: 0)
+                return self.compositionalLayout.enrollLayout(
+                    portraitHorizontalNumber: 1,
+                    landscapeHorizontalNumber: 1,
+                    cellVerticalSize: .fractionalHeight((1 - (1/6))/CGFloat(self.placeholderList.count)),
+                    scrollDirection: .vertical,
+                    cellMargin: listCellMargin,
+                    viewMargin: listViewMargin)
+            }
```
조건으로 들어오는게 현재 2개이면 if 조건문만으로 충분할것 같다고 생각했는데 switch로도 접근할 수 있겠네요!?
산체스는 어떨때 switch 로직을 사용하는 기준이 있나요🧐

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+    
+    func PassListCellData(
```
함수명이 대문자로 시작합니다😱

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,130 @@
+        return textFieldDataPass
+    }
+    
+    func CheckCellData(
```
함수명이 대문자로 시작합니다😱

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDelegate.swift:
```diff
@@ -0,0 +1,33 @@
+                    collectionView.indexPathsForSelectedItems else { return }
+            for indexPath in photoIndexPaths {
+                enrollModifyCollectionViewDataSource?.photoAlbumImages.remove(
+                    at: indexPath.item - 1)
```
파라미터가 하나일때는 한라인에 표현하는것이 더 가독성이 좋을것 같다고 생각드는데 산체스는 어떤가요??

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarket/OpenMarketCollectionViewDataSource.swift:
```diff
@@ -57,21 +58,23 @@ extension OpenMarketCollectionViewDataSource: UICollectionViewDataSource {
+                                       landscapeHorizontalNumber: 1,
+                                       cellVerticalSize: .absolute(100),
+                                       scrollDirection: .vertical,
+                                       cellMargin: nil, viewMargin: listViewMargin)
```
코드 컨벤션을 위해 viewMargin도 개행을 하여 선언해주면 좋을것 같습니다😀

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/EnrollModify/EnrollModifyCollectionViewDelegate.swift:
```diff
@@ -0,0 +1,33 @@
+            for indexPath in photoIndexPaths {
+                enrollModifyCollectionViewDataSource?.photoAlbumImages.remove(
+                    at: indexPath.item - 1)
+                enrollModifyCollectionViewDataSource?.medias.remove(
```
이렇게 사용되면 혹시 등록된 사진을 'x' 버튼을 통해 삭제하려고하면 인덱스아웃 크래쉬가 나지 않을까요❓❓

GREENOVER commented on OpenMarket/OpenMarket/Controller/EnrollModifyViewController.swift:
```diff
@@ -0,0 +1,190 @@
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        
+        delegate?.changeOrientation = true
```
요 부분의 코드는 어떤 의미있는 동작을 하는지 궁금합니다🙋🏻‍♂️
상품등록/취소 뷰는 화면 전환을 일으키지 않으려는것 같은데 
landscape로 진입하면 landscape로 나오고
portrait으로 진입하면 portrait으로 고정되는것이 의도여서 좋은것 같은데
두 방향에 대한 뷰를 지원한다면 화면 회전 전환을 살려도 괜찮지 않을까요??
아니면 portrait으로만 가져가는것도 좋아보입니다😀

GREENOVER commented:
안녕하세요 산체스!
스텝3까지 오시느라 정말정말 고생많으셨습니다🎉🎉🎉
비록 스텝3를 다하지 못하셨어도 충분히 그 속에서 많은 구현과 시도를 해보며 더 발전하신게 훤히 보이네요👀
저도 캠퍼때 스텝을 다 못끝내거나 계획했던 부분까지 완수를 못하면 계속 생각나고 잠도 안오고 했는데
산체스도 그런 열정을 갖고 있는것 같아서 3주간 정말 리뷰할 맛 났습니다!
이번 리뷰는 스텝3에서 산체스가 주로 구현한 뷰 구성과 포토쪽을 주력으로 말해볼 것 같아요!
아무래도 다른 부분들은 거의 되셨던것 같은데 포토 등록쪽과 여타 부분들에서 아주 조그마한 문제로 꼬여서
시간을 많이 잡아먹은것 같아 보였습니다😱
그래도 정말 코드 퀄이나 스텝1 처음의 산체스를 스스로 비교해봤을때 많은 성장을 했구나 느꼈을것 같습니다👍🏻
다시 한번 수고 많으셨어요👏👏👏

***
### 산체스를 칭찬합니다👏
- 코드와 파일을 역할별로 분리하는 능력이 이제는 수준급이네요💯
- 커스텀한 뷰를 그리느라 고생많으셨을텐데 오토레이아웃도 그렇고 흠이 없습니다💯
- 네이밍에 대해 이제는 산체스의 생각이 잘 드러나고 이전보다 훨씬 명확해져서 가독성이 좋아요💯
- 안되는 부분도 포기하지 않고 어떻게든 해결하려는 열정💯

### 궁금한점❓
- 이번에도 타이틀로 걸만한 궁금한 점은 딱 정할 수 없었지만 PR에 어느정도까지 구현을 한건지 명시가 되있으면 더 좋았을것 같습니다😊
- 등록/수정 화면을 하나의 CollectionView로 만든다는 생각을 저는 못했을것 같은데 그런 방법으로 접근을 하셨던게 정말 돋보입니다👍🏻
혹시 역할이 다르고 어떻게 보면 아주아주 작은 차이로 UI의 구성도 다른 두 뷰를 CollectionView로 한 뷰로 묶은 이유가 궁금합니다🙋🏻‍♂️
(어떤 이점을 가져가려했는지에 대해 궁금해요ㅎㅎ)
- 스텝3는 등록/수정이라는 아예 산체스만의 커스텀한 뷰를 제작해보는 과정이였는데 UI 구성 및 디자인은 어떤 과정으로 현재의 결과물로 도출되었는지 그 절차가 궁금합니다🙋🏻‍♂️
> 더 나아가보면 애플의 가장 큰 장점중 하나라 생각되는 부분은 '사용자 경험'이라고 저는 생각합니다. 
> 애플이 그 부분을 기가맥히게 잘 캐치하고 신경쓰거든요. 
> 그래서 저는 좋은 앱의 기준 중 하나를 그 부분을 잘 녹였고 고려했는지로 판단합니다. 그런 측면에서 UI를 구성할때 참고가 될 수 있을거에요!
> https://developer.apple.com/kr/design/tips/
> 이런 자료들도 심심하실때 한번씩 보시면 애플의 사고를 이해할 수 있을겁니다👍🏻

### 산체스의 고민에 대한 저의 생각🤔
- 포토앨범과 사진을 가져오는 부분에 대해 많은 고민과 노력이 있었던것이 코드에서 많이 보였어요👍🏻
저는 지금 산체스의 방향 좋다고 생각합니다. 당연히 라이브러리를 사용하면 간단하지만 그렇지 않다는 가정하에 산체스처럼 커스텀하게 뷰와 로직을 구성하고 구현해보는 이번 과정이 정말 값지다고 생각해요. 산체스가 포토앨범 멀티픽 기능을 구현하시려는 부분에서 기능과 코드적으로 크게 잘못된 방향으로 간다는 점은 찾지 못했습니다. 지금 방향도 좋다는 의미입니다ㅎㅎ 다만, 아직 구현이 덜 되었는지 아니면 커스텀 구현의 한계를 느끼셨던건지 잘 모르겠어서.. 의문을 제기해봅니다☺️
앨범으로 접근을 하고 사진을 픽하면 커스텀한 포토앨범에 사진들의 이미지가 빈 이미지로 보이는 부분이 있어요🥲
사진을 멀티픽한 만큼 임포트해온것 같긴한데 사진이 아예 보이지 않는 문제에 대해 보완되면 이상없을것 같습니다.
또 그 보이지 않는 사진들을 체크하고 확인을 눌렀을때 상품등록/조회 뷰에 해당 사진들이 등록은 되는것은 확인되지만 사진 수를 나타내는 0/5의 텍스트가 변하지 않는 짜잘한 부분들의 로직이 빠져있어서 간단히 추가하면 되겠다 생각했어요!
그리고, 해당 사진들을 X 버튼으로 삭제할때는 크래쉬가 발생하는 코드가 보여 해당 부분 리뷰로 옵저버 심어두었습니다😁
이런 사소한 부분들의 추가 구현만 이뤄진다면 산체스의 현재 접근한 방법도 좋다고 보입니다👍🏻
- 등록/수정화면을 한 뷰로 묶어 처리하면서 많은 부분을 잡아먹으신것 같아요🥲
어떤 의도인지 조금은 이해는 되지만 저는 오버엔지니어링이란 생각도 듭니다..!
산체스가 제가 말한 응집도/결합성에 관한 이야기때문에 최대한 공통분모를 묶어 처리하고 코드를 줄이려하는건가 싶어 걱정도 되었어요!
그래서 뷰를 두개로 따로 가져가면 더 가독성도 좋고 구현이나 추후 리팩토링할때도 용이하지 않을까 싶었습니다.
제가 이해한 산체스의 의도는 등록/수정이 어떻게보면 공통된 뷰를 지니고 있어 공통적으로 처리하는게 중복도 줄이지 않을까라고 이해했는데 맞을까요?
그런 관점에서 저는 생각해보자면 이런 경우는 충분히 등록과 수정이 각각 기획이 변할 수도 있다고 봐요.
그럴때는 공통적으로 묶었기에 리팩토링하기가 까다로워질 수 있습니다.
두 파트는 분리해서 생각하고 가져가는것이 저는 개인적으로 나아보인다고 생각해요🙌
그렇지만 굉장히 좋은 시도이고 충분히 산체스가 의도한바 대로 나왔다면 오버엔지니어링이 아닌 스마트한 코드였을겁니다👍🏻
- 등록/수정 부분에 대한 프로토콜 처리와 고민들도 위의 의견과 같은 맥락이라고 생각합니다🤓
산체스의 프로토콜 활용 능력은 정말 일품이라고 프로젝트 내내 생각했습니다.
다만 오버엔지니어링이 되지 않도록 노력하는것과 동시에 모든것을 하나의 틀의 사고로 생각하지 않도록 경계해야합니다🧐
- 뷰의 기능을 따라 나누는 고민을 쭉 많이 이번 프로젝트에서 해오셨는데 정말 많은 성과가 있었다고 저는 생각합니다!
한 뷰에서 다 처리를 해주는것이 틀린건 아니에요. 개발할때도 어떻게 보면 방대한 코드가 아니라면 편할 수도 있고요.
그렇지만 그렇게 하지 않고 분리하는 이유와 장점에 대해서는 산체스도 많이 느껴 이러한 좋은 코드가 나왔을거라 생각합니다👍🏻
물론 짜치고 시간이 가지만 개발 다음엔 리팩토링에 연속이잖아요? 그럴때 드는 유지보수 시간 비용은 훠얼씬 줄어들거에요ㅎㅎ
지금 산체스가 하는 방식이 옳다고 봅니다🙋🏻‍♂️
다만 하나의 기술에 의존하는것을 경계하면서 항상!!ㅎㅎ
이번에도 뷰를 나누고 네비게이션컨트롤을 적재적소에 잘 해주었던 부분도 인상깊어요👍🏻
***
이번 리뷰로 산체스에게 의문을 좀 더 던져준것 같아 죄송스럽지만 명절이니까 즐겨보자구요😃
3주간 산체스를 리뷰할 수 있어서 저에게도 너무 큰 영광이였습니다.
많은 부분에서 배우고 자극 받기도 했어요⚡️
3주간 일정에 대해 공유하고 조율해주셔서 저또한 일정 관리 하는데 편했습니다☺️
그런 사이드적인 능력도 충분한 장점입니다. 
현업에서 꼭꼭 필요해요👍🏻
산체스는 코딩 스킬뿐 아니라 그런 다양한 장점들도 탑재한 만능 M1 같은 존재이니 어떤 프로젝트던 잘 완수하실 수 있다고 봅니다🤩
다시 한번 3주간 너무너무 고생많으셨고 감사합니다🙇🏻
추석 명절겸 캠퍼 방학일텐데 맛있는거 많이 드시고 체력 보충해서 푹쉬고 오세요!
(방학때 많이 공부하지 마세요.. 쉴 수 있을때 캠프도 쉬어야 합니다! 경..험..담..😭)
+ 언제든 궁금하신거나 요 플젝에 대해 더 진행하고 싶거나 리뷰 받고 싶으시면 개인적으로도 PR 주셔도 되요😁

sanches37 committed:
refactor: select photos 선택시 오류 수정
사진첩 접근시 select photos 를 선택하면 불러오는 시점 문제 수정
기존에 UIImage 로 collectionView 에서 사용했는데 asset 에서 사용으로 변경

sanches37 committed:
refactor: CheckCellData 사용하지 않도록 코드 수정
코드를 줄이기 위해 CheckCellData 없애고 PassListCellData 로 필수항목 파악

sanches37 commented:
안녕하세요 그린. 리뷰를 빠르게 해주셨는데 답변을 이제서야 달게되었네요.
포토앨범 선택시 select photos 를 선택했을 때 오류가 나는 부분을 확인하지 못했었네요. 첫 부분부터 큰 오류가 나서 리뷰하실 때 보실게 없으셨을 것 같습니다. 이 부분은 불러오는 시점의 문제인 것 같아서 수정을 진행했습니다.

* 어디까지 구현했는지를 제대로 말씀드리지 않았군요. 이미지를 선택해서 텍스트필드에 내용을 기입하고 등록이나 수정을 할 수 있는 부분까지는 구현했지만 그 과정에서 이미지 선택시 몇장의 이미지를 선택했는지는 notification으로 전달해서 구현하고 싶었는데 시간이 부족했습니다. 지금 코드에서는 textfield 설정도 아에 되어있지 않아서 키보드 관련 부분 코드를 추가해야 될 것 같아요. 특히 가격이나 재고 , 할인가격은 숫자 키보드만 사용되게 적용해야 될 것 같습니다. 등록/수정시 필수사항들이 입력되지 않으면 등록/수정이 되지 않는 로직을 어느 정도 구현은 했는데 등록시 image를 무조건 선택해야되는데 그부분에 대한 확인 로직은 추가하지 못했습니다. 추가로 등록/수정 화면을 세로모드로만 보이게 하려고 했는데, 이전화면에서 가로모드로 진입했을 때 등록/수정화면도 가로모드로 보이게 되니까 그 부분도 수정해야 될 것 같습니다.

* 제가 구현하려는 방식에서는 사진 선택을 가로 스크롤을 사용해야되서 collectionView 를 사용하는건 고민을 많이 안했었어요. 고민한 부분은 스택뷰를 이용한 textField 를 만들지 tableView 나 collectionView 를 이용한 textField 를 만들지 였는데, 구축엔 조금 어려움이 있지만 항목이 추가되거나 삭제 됐을 때 배열의 count 수로 화면에 고정적으로 표시할 수 있어서 변화대응에 좀 더 좋지 않을까해서 tableView 또는 collectionView 로 구현해보려 했어요. 구현하던 중 textFiled 를 collectionView 로 선택하면 section 구분만 해주면 collectionView를 굳이 2개를 사용하지 않아도 될 것 같아서 이러한 방식으로 구현하게 됐습니다.

UI 구성 및 디자인은 당근마켓의 중고거래 등록 페이지를 참고했었어요. 이대로 한번 따라 만들어보자는 생각으로 접근해서 만들게 됐습니다.
<img src="https://user-images.githubusercontent.com/84059338/133917713-5b2b6e7c-86d7-4f92-86a1-e84f656bfae9.PNG"  width="350" height="600"/>

* 말씀하신대로 등록/수정화면은 어떤식으로 바뀔지 알 수 없는 상황에서 뷰를 두개로 나눠서 관리하는게 저도 좋을 것 같습니다!
이 부분은 프로젝트 핵심경험의 상속 혹은 프로토콜 기본구현을 통해 (수정/등록 과정의) 공통기능 구현을 보고 뷰를 하나로 쓰고 프로토콜로 선택할 수 있게 구현해보라는 의미구나 생각하고 지금의 구현을 하게되었는데, 프로토콜을 쓰더라도 뷰는 두개로 나눠서 할 수 있을건데 생각의 폭이 좁았던 것 같습니다.

스탭3을 완성하기 위해 추가해야 될 사항들이 많이 남았고, 리뷰해주신 부분도 수정을 해야 될 것 같아요. 3주간 조금 빡세게 달렸는지 말씀하신대로 방학은 조금 쉬어야 될 것 같습니다. ㅎㅎ
방학 동안 쉬면서 이번 프로젝트 리드미를 작성을 해보려합니다. 스탭 완성과 리뷰수정은 캠프가 끝나면 다시 한번 이어서 진행해보겠습니다.
3주동안 리뷰해주셔서 정말 감사합니다.

GREENOVER reviewed (approved):
@sanches37 
고생 많으셨습니다 산체스!
해당 포토앨범의 동작은 군더더기 없이 잘 동작하는걸 확인하였습니다👍🏻
스텝3에서 나머지 기능들의 구현이 미비한 부분은 있지만 우선 머지하겠습니다!
이후 다른 프로젝트들을 하느라 시간이 없으실수 있지만 추후에 포폴로 다시 완성해보거나 할때
언제든 여쭤보시면 도와드리겠습니다!
정말 3주동안 고생 많으셨고 너무 잘해주셔서 리뷰하기도 편했고 많이 배웠습니다🙇🏻
방학 푹 쉬고 재충전해서 다음주부터 또 달려봐요!🙏🏻
